File: anim.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# TODO: All of this assumes a framerate of 30!

def animationStep(nodeName, attrName, startValue, endValue, 
        duration, curTime):
    node = g_Player.getElementByID(nodeName)
    if (curTime < duration): 
        curValue = startValue+(endValue-startValue)*curTime/duration
        setattr(node, attrName, curValue)
        g_Player.setTimeout(30,
                lambda: animationStep(nodeName, attrName, startValue,
                        endValue, duration, (curTime+30)))
    else:
        setattr(node, attrName, endValue)

def animateAttr(Player, nodeName, attrName, startValue, endValue, duration):
    global g_Player
    g_Player = Player
    animationStep(nodeName, attrName, startValue, endValue, duration, 30)

def fadeStep(nodeName, change): 
    node = g_Player.getElementByID(nodeName)
    node.opacity += change

def fadeEnd(id, nodeName, val):
    node = g_Player.getElementByID(nodeName)
    node.opacity = val
    g_Player.clearInterval(id)

def fadeOut(Player, nodeName, duration):
    global g_Player
    g_Player = Player
    node = g_Player.getElementByID(nodeName)
    durationInFrames = duration*30/1000
    changePerFrame = -node.opacity/durationInFrames
    id = g_Player.setInterval(25, lambda: fadeStep(nodeName, changePerFrame))
    g_Player.setTimeout(duration, lambda: fadeEnd(id, nodeName, 0))

def fadeIn(Player, nodeName, duration, max):
    global g_Player
    g_Player = Player
    node = Player.getElementByID(nodeName)
    durationInFrames = duration*30/1000
    changePerFrame = (max-node.opacity)/durationInFrames
    id = Player.setInterval(25, lambda: fadeStep(nodeName, changePerFrame))
    Player.setTimeout(duration, lambda: fadeEnd(id, nodeName, max))

File: anim.js
// TODO: All of this assumes a framerate of 30!

function animationStep(nodeName, attrName, startValue, endValue, duration, 
curTime)
{
    var node = AVGPlayer.getElementByID(nodeName);
    if (curTime < duration) {
        var curValue = startValue+(endValue-startValue)*curTime/duration;
        eval("node."+attrName+"="+curValue);
        var code = "animationStep(\""+nodeName+"\", \""+attrName+"\", "
                +startValue+", "+endValue+", "+duration+", "+(curTime+30)+");";
        AVGPlayer.setTimeout(30, code);
    } else {
        eval("node."+attrName+"="+endValue);
    }
}

function animateAttr(nodeName, attrName, startValue, endValue, duration)
{
    var code = "animationStep(\""+nodeName+"\", \""+attrName+"\", "
                +startValue+", "+endValue+", "+duration+", "+30+");";
    eval(code);
}

function fadeStep(nodeName, change) 
{
    var node = AVGPlayer.getElementByID(nodeName);
    node.opacity += change;
}

function fadeEnd(id, nodeName, val)
{
    var node = AVGPlayer.getElementByID(nodeName);
    node.opacity = val;
    AVGPlayer.clearInterval(id);
}

function fadeOut(nodeName, duration)
{
    var node = AVGPlayer.getElementByID(nodeName);
    var durationInFrames = duration*30/1000;
    var changePerFrame = -node.opacity/durationInFrames;
    var id = AVGPlayer.setInterval(25, 
            "fadeStep(\""+nodeName+"\","+changePerFrame+");");
    var code = "fadeEnd(\""+id+"\",\""+nodeName+"\", 0);";
    AVGPlayer.setTimeout(duration, code);
}

function fadeIn(nodeName, duration, max)
{
    var node = AVGPlayer.getElementByID(nodeName);
    var durationInFrames = duration*30/1000;
    var changePerFrame = (max-node.opacity)/durationInFrames;
    var id = AVGPlayer.setInterval(25, 
            "fadeStep(\""+nodeName+"\","+changePerFrame+");");
    var code = "fadeEnd(\""+id+"\",\""+nodeName+"\", "+max+");";
    AVGPlayer.setTimeout(duration, code);
}

